org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/resource/FileStoreStorage.java

```diff
@@ -0,0 +1,80 @@
+/*******************************************************************************
+ * Copyright (c) 2018 itemis AG (http://www.itemis.de) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ui.resource;
+
+import java.io.InputStream;
+
+import org.eclipse.core.filesystem.EFS;
+import org.eclipse.core.filesystem.IFileInfo;
+import org.eclipse.core.filesystem.IFileStore;
+import org.eclipse.core.resources.IEncodedStorage;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+
+/**
+ * {@link IEncodedStorage} implementation for external files. Used to support open reference for external files.
+ * 
+ * @author Arne Deutsch - Initial contribution and API
+ */
+public class FileStoreStorage implements IEncodedStorage {
+	private final IFileStore fileStore;
+	private final IFileInfo fileInfo;
+	private final Path filePath;
+
+	public FileStoreStorage(IFileStore fileStore, IFileInfo fileInfo, Path filePath) {
+		this.fileStore = fileStore;
+		this.fileInfo = fileInfo;
+		this.filePath = filePath;
+	}
+
+	@Override
+	public <T> T getAdapter(Class<T> adapter) {
+		return null;
+	}
+
+	@Override
+	public InputStream getContents() throws CoreException {
+		return fileStore.openInputStream(EFS.NONE, null);
+	}
+
+	@Override
+	public IPath getFullPath() {
+		return filePath;
+	}
+
+	@Override
+	public String getName() {
+		return filePath.lastSegment();
+	}
+
+	@Override
+	public boolean isReadOnly() {
+		return fileInfo.getAttribute(EFS.ATTRIBUTE_READ_ONLY);
+	}
+
+	public IFileStore getFileStore() {
+		return fileStore;
+	}
+
+	@Override
+	public String getCharset() throws CoreException {
+		return null; // encoding unknown, use default encoding
```

Comment:
Why do we implement IEncodedStorage if we do nothing about the encoding?

Reply:
The code using the instance makes a "instanceof" check. If the class does not implement "IEncodedStorage" it falls back to ask the workspace for the encoding. Leading to an exception in the end. Implementing the interface leading to a null. In turn the "default" encoing from eclipse is used what is the bast we can do at this place (as the encoding of external files is unknown).

Reply:
Ok